### ISSUE TITLE
Optionally produce component which imports libpython311.so, libc.so, etc. instead of bundling them

### ISSUE THREAD

**Issue by dicej** [bytecodealliance/componentize-py#28]
This would dramatically reduce binary sizes in cases where the host can provide these libraries.

**Comment by dicej** [comment]
I did some research to determine a size breakdown for a "hello world" wasi-http component generated by `componentize-py`.

The TL;DR is that the stripped binary is about 18MB, 12MB of which is the data section with the pre-init memory snapshot, and the rest of which are the various modules, the largest of which is libpython312.so at about 4MB.

A stock, stripped libpython312.so is about 8MB, 4MB of which is data section(s) which get replaced with the memory snapshot when we pre-init.

The upshot is that we could save about 6MB per component by importing the modules it needs, but we'd still be looking at 12MB minimum component sizes.  To reduce sizes further, we'd need some way to package the `.py`, `.pyc`, and related standard library files into a composable VFS component we could publish to a registry and reuse.  And we'd also need a standard way to export a pre-init function that can be run by the host (e.g. at deployment time) to minimize cold-start times since we wouldn't be doing pre-init at build time anymore.  See also https://github.com/bytecodealliance/componentize-py/issues/23

**Comment by QuantumSegfault** [comment]
Any word on this? This is something of a blocker for us over at @Pumpkin-MC.

`componentize-py` works, but the resulting binaries are large enough that it's impractical to use them. We have a 5 MiB size limit for plugins on our [market](https://market.pumpkinmc.org/), But minimum size is 15+ MiB. And (at least with debug builds; maybe it's not nearly as bad on proper release builds) this large size significantly increases the time it takes for Wasmtime to cold-start the plugin (make and cache the `.cwasm`, etc.; we're talking minutes).

We're fine with a little extra work to ship the CPython artifacts with Pumpkin (e.g. it doesn't NEED to be automated via `warg`/`wpkg`, though it could be nice). But that requires being able to split the user plugin code, from everything else.

-----

I imagine that the way things would work is that our tool would bundle the user's Python source into a VFS, and have a little bit of glue that on plugin startup, runs starts the interpreter (using an import from the Python component, supplied by the host to the component linker) using the code in the plugin's VFS (and I guess that same interpreter component would have to include the stdlib VFS as well).

Not sure how component model imports/exports fit into the picture though with it fragmented like that...

Probably also going to need `wac` in order to be able to rewire the plugin/component to inject the imported interpreter and Python stdlib VFS in the right spot (though it would make for a good practical example of the composability of the component model in action).

**Comment by dicej** [comment]
Hi @QuantumSegfault.  I'm not aware of anyone actively working on this, nor do I have any near-term plans to do so myself.  I do agree that it's a real problem for many use cases, and it sounds like yours is one of them.

Expanding on the above discussion, here's a sketch of what I would consider a minimum viable solution to this problem:

- `componentize-py` support for emitting components that _import_ `libpython3xx.so`, `libc.so`, `libc++.so`, `libcomponentize_py_runtime_async.so`, etc. instead of bundling them.
    - This would also include importing a _component_ (i.e. not just a `.so` module) containing a VFS with the stdlib.  In other words, the emitted component would include only a small stub module that links all the imported modules and component together and hooks them up to one or more component-level exports, plus an internal VFS containing the plugin's code (as you described), and possibly some additional, plugin-specific `.so` files (e.g. native extensions)
    - Finally, `componentize-py` would need to skip the pre-init step and instead ensure that any initialization code is run lazily the first time an exported function is called (or during component instantiation)
- `wasmtime` support for lazily generating and caching separate `.cwasm` files for each module and/or component imported by a component.  Based on previous conversations with @alexcrichton, this should mostly "just work", but may or may not require some extra configuration by the embedding application.

The two biggest areas of uncertainty for me are:

-  Constructing and layering the VFS stack (plugin VFS, stdlib VFS, and runtime FS).  Presumably we'd use `wasi-virt` for this, but I haven't tried using that for anything serious yet.  Hopefully it just works?
- Running components in Wasmtime which import modules and/or subcomponents.  Again, hopefully it just works?

Again, this isn't a high priority for me, but I'd be happy to mentor anyone who'd like to tackle it.

**Comment by alexcrichton** [comment]
One thing I can note is that components-importing-modules works as-is today in Wasmtime, albeit there's not a ton of stress on that part of Wasmtime so bugs may be lurking there. Components-importing-components is not supported in Wasmtime adn more-or-less can't be supported without a pretty large amount of investment in new work. For components-importing-components what'd be required is that the host would have to compose-on-the-fly to present a component to Wasmtime that doesn't import other components anywhere. Internally components can import each other and that should work out (modulo not being stressed much so may have slight bugs), but it's the outer component that's presented to Wasmtime which can't import a component.

I'll also note that none of this can be exercised on the `wasmtime` CLI at this time for testing and such. That's a feature we should add though! (and wouldn't be too hard, so long as it's just saying "hey provide this module to the component I'm loading" as opposed to compose-on-the-fly)

**Comment by dicej** [comment]
OK, so we will need to use `wac` for the VFS layering as @QuantumSegfault suggested, at least for the time being, or else assume the application (e.g. Pumpkin) will provide the stdlib files via a host:guest mount at runtime.

**Comment by QuantumSegfault** [comment]
Huh... I didn't know you could import Wasm modules with components. I would have imagined we would bundle the entire Python runtime (`so` and all) into a self-contained component that supplies a single export ("run this Python file from the FS") imported and invoked by the final component.

Is it Python native extensions that are making this weird?

------

Really...components importing other components doesn't work? So only the host can provide implementations of required interfaces directly? You can't load up a bunch of different components, provide them to the Wasmtime component linker, and let other component exports resolve satisfy some imports?

**Comment by alexcrichton** [comment]
> You can't load up a bunch of different components, provide them to the Wasmtime component linker, and let other component exports resolve satisfy some imports?

Correct, yeah. Nothing on [`component::Linker`](https://docs.rs/wasmtime/latest/wasmtime/component/struct.Linker.html) provides the ability to say "here's a component" or something like "here's a function from an instantiated component". Note, though, that a [module can be provided](https://docs.rs/wasmtime/latest/wasmtime/component/struct.LinkerInstance.html#method.module).

Instantiating one component with another is pretty tricky because you have to first figure out how to instantiate the first one, then hook that up into the second one the way it wants to slot in. It's not quite so simple (for now) as just plugging two things together and calling it a day. For example `wac` is a pretty nontrivial project and that's what does the "plugging". Note, though, that you can always `wac` a component together and present that to Wasmtime.

**Comment by dicej** [comment]
> I would have imagined we would bundle the entire Python runtime (`so` and all) into a self-contained component that supplies a single export ("run this Python file from the FS") imported and invoked by the final component.

Yes, you can do that today for pure Python code, if you like.  You can use `componentize-py` to create a "plugin-runner" component that either targets some custom "plugin" WIT world you specify or just the standard `wasi:cli/command` world, based on a Python program that runs arbitrary Python code, similar to [this example](https://github.com/bytecodealliance/componentize-py/blob/main/examples/sandbox/guest.py).

In that scenario, Pumpkin plugins written in Python might just be tarballs of Python source code to be fed to the "plugin-runner" component.

> Is it Python native extensions that are making this weird?

Indeed, that's where things get awkward.  Currently, there's no way in the component model to load modules on-the-fly at runtime; they need to be baked into the component.  See [this](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Linking.md) and especially [this](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Linking.md#fully-runtime-dynamic-linking) for details.

**Comment by QuantumSegfault** [comment]
!

Okay...then I might be able to make this work without upstream changes...

I'll have to investigate this approach.

I think we'd still package it using `wasi-virt`. Thanks for the tip!

**Comment by dicej** [comment]
I just realized we'll probably need to address https://github.com/bytecodealliance/componentize-py/issues/23 to make that approach work in the general case (e.g. when the plugin needs to use parts of the stdlib not pulled in during pre-init by the plugin runner).
